docs/guides/intro-to-tests.md

```diff
@@ -0,0 +1,154 @@
+
+### Case 1 🖊️
+
+```js
+const input = "hello";
+const result = createWordCount(input);
+```
+
+In example 1, `input` stores a string `"hello"` which we pass as an argument to `createWordCount`.
+We then store the return value of `createWordCount` in a variable called `output`.
+In this case, we would expect the `output` to be:
+
+```js
+{ "hello" : 1 }
+```
+
+Remember that `createWordCount` needs to return an object with a count of each word in the string. In the case where the input is `"hello"` we only have a single word string and therefore a count of 1 for `"hello"`.
+
+### Case 2 🖊️
+
+```js
+const input = "north by north west";
+const result = createWordCount(input);
+```
+
+In example 2, `input` stores a string `"north by north west"` which we pass as an argument to `createWordCount`.
+We then store the return value of `createWordCount` in a variable called `output`.
+In this case, we would expect the `output` to be:
+
+```js
+{
+    "north" : 2,
+    "by" : 1,
+    "west" : 1
+}
+```
+
+## 🧪 What is a test ?
+
+We can now define what a test is in software engineering.
+
+> 🔑 A **test** is any piece of code that **checks a certain unit of code** ( typically a function ) **behaves in a particular way**.
+
+For example, we could write a test to check that `createWordCount` behaves in a particular way when passed a string containing a single word. For this example, we're using a popular JavaScript testing framework called [Jest](https://jestjs.io/docs/getting-started) to write the test code. We'll examine the example below written using Jest, and then break down the different parts of the test code to understand how it all works.
+
+```js
+test("will count the number of words for a single word string", function () {
+  const input = "hello";
+  const actualResult = createWordCount(input);
+  const expectedResult = {
+    hello: 1,
+  };
+  expect(actualResult).toEqual(expectedResult);
+});
+```
+
+Let's break down the different parts of this test code.
+
+## Anatomy of a test
+
+#### 1. Test description
+
+```js
+test("will count the number of words in a single word string", function () {});
+```
+
+We use a function in Jest called `test` to write a test description - `"will count the number of words for a single word string"` which describes the behaviour we expect in this particular test. There is also a function `function () {}` in which we will write our test code.
+
+#### 2. Test setup
+
+```js
+const input = "hello";
+const actualResult = createWordCount(input);
+const expectedResult = {
+  hello: 1,
+};
+```
+
+In this section, we're declaring some important values that we will need to compare later on in the test.
```

Comment:
I can see some of our trainees reading this as a required recipe (you need a variable named `input` and a variable named `actualResult`, etc) - while the note after is helpful, it may also be worth adding an introductory sentence/paragraph before this saying something like "We can write any code we want in the test function. The important thing is that we have at least one assertion, but we can use all of the techniques we learn in javascript (like using variables, or calling functions) to help us to write our test."

Which actually makes me thing... Maybe we can build this up the other way around? Start with a test that looks like:

```js
test("will count the number of words for a single word string", function () {
    expect(createWordCount("hello")).toEqual({hello: 1});
});
```

so that we can quickly explain assertions, and then break apart the test to use variables and show that it can make it more clear?

Reply:
In my experience, extracting variables out of a complex example can sometimes lose trainees when they're not confident with expressions/statements. On the other hand, we hopefully should have addressed that so maybe it's not a concern.

I do agree that showing a inlined example is probably worth doing. It's common enough practice that trainees looking for more materials online will see it, so I think it's worth showing and discussing as part of the class.